nestkernel/conn_builder.cpp

```diff
@@ -602,34 +609,78 @@ nest::OneToOneBuilder::connect_()
       // allocate pointer to thread specific random generator
       librandom::RngPtr rng = kernel().rng_manager.get_rng( tid );
 
-      for ( GIDCollection::const_iterator tgid = targets_->begin(),
-                                          sgid = sources_->begin();
-            tgid != targets_->end();
-            ++tgid, ++sgid )
+      // With large-scale simulations (vp >> indegree), most of the nodes in the
+      // targets_-range are not on this machine. Therefore, if the number of
+      // local nodes is smaller than the targets_-range, we will iterate over
+      // the local nodes instead and check, whether the node is in the
+      // targets_-range.
+      if ( targets_->size() < get_local_nodes().size()
+        || not targets_->is_range()
+        || parameters_requiring_skipping_.size() > 0 )
       {
-        assert( sgid != sources_->end() );
+        for ( GIDCollection::const_iterator tgid = targets_->begin(),
+                                            sgid = sources_->begin();
+              tgid != targets_->end();
+              ++tgid, ++sgid )
+        {
+          assert( sgid != sources_->end() );
 
-        if ( *sgid == *tgid and not autapses_ )
-          continue;
+          if ( *sgid == *tgid and not autapses_ )
+            continue;
 
-        // check whether the target is on this mpi machine
-        if ( not kernel().node_manager.is_local_gid( *tgid ) )
-        {
-          skip_conn_parameter_( tid );
-          continue;
-        }
+          // check whether the target is on this mpi machine
+          if ( not kernel().node_manager.is_local_gid( *tgid ) )
+          {
+            skip_conn_parameter_( tid );
+            continue;
+          }
 
-        Node* const target = kernel().node_manager.get_node( *tgid, tid );
-        const thread target_thread = target->get_thread();
+          Node* const target = kernel().node_manager.get_node( *tgid, tid );
+          const thread target_thread = target->get_thread();
 
-        // check whether the target is on our thread
-        if ( tid != target_thread )
-        {
-          skip_conn_parameter_( tid );
-          continue;
+          // check whether the target is on our thread
+          if ( tid != target_thread )
+          {
+            skip_conn_parameter_( tid );
+            continue;
+          }
+
+          single_connect_( *sgid, *target, target_thread, rng );
         }
+      }
+      else
+      {
+        for ( SparseNodeArray::const_iterator it = get_local_nodes().begin();
```

Comment:
@jakobj The problem with a `GidCollection` would be that it represents gobal sets of GIDs in a rather abstract fashion that first needs to be expanded. The `SparseNodeArray` gives us access to local node pointers right away. But we are risking encapsulation here, so I will look at how to implement this cleaner.